// htmlhint/dist/cli/formatters/sarif.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const node_sarif_builder_1 = require("node-sarif-builder");
const path = require("path");
const pkg = require('../../../package.json');
function getRuleMarkdown(ruleId) {
    const possiblePaths = [
        path.join(process.cwd(), 'website', 'src', 'content', 'docs', 'rules', `${ruleId}.mdx`),
        path.join(path.dirname(require.resolve('../../../package.json')), 'website', 'src', 'content', 'docs', 'rules', `${ruleId}.mdx`),
        path.join(process.cwd(), 'src', 'content', 'docs', 'rules', `${ruleId}.mdx`),
    ];
    for (const mdxFilePath of possiblePaths) {
        try {
            if ((0, fs_1.existsSync)(mdxFilePath)) {
                const content = (0, fs_1.readFileSync)(mdxFilePath, 'utf8');
                const frontmatterEnd = content.indexOf('---', 4) + 3;
                if (frontmatterEnd > 3) {
                    const markdown = content.substring(frontmatterEnd).trim();
                    const lines = markdown.split(/\r?\n/);
                    const filteredLines = lines.filter((line) => !line.includes("import { Badge } from '@astrojs/starlight/components';"));
                    let processedMarkdown = filteredLines.join('\n');
                    processedMarkdown = processedMarkdown.replace(/<Badge\s+text="([^"]+)"[^>]*\/>/g, '$1');
                    processedMarkdown = processedMarkdown.replace(/(<\/?[a-zA-Z][^>\s]*[^>]*>|<!DOCTYPE[^>]*>)/g, '`$1`');
                    processedMarkdown = processedMarkdown.replace(/```html\n([\s\S]*?)```/g, (match, code) => `\`\`\`html\n${code}\`\`\``);
                    return processedMarkdown;
                }
            }
        }
        catch (error) {
        }
    }
    return undefined;
}
const sarifFormatter = function (formatter) {
    formatter.on('end', (event) => {
        const arrAllMessages = event.arrAllMessages;
        const sarifBuilder = new node_sarif_builder_1.SarifBuilder();
        const sarifRunBuilder = new node_sarif_builder_1.SarifRunBuilder().initSimple({
            toolDriverName: 'HTMLHint',
            toolDriverVersion: pkg.version,
            url: 'https://htmlhint.com/',
        });
        const addedRuleSet = new Set();
        arrAllMessages.forEach((result) => {
            result.messages.forEach((message) => {
                const rule = message.rule;
                if (addedRuleSet.has(rule.id)) {
                    return;
                }
                addedRuleSet.add(rule.id);
                const sarifRuleBuilder = new node_sarif_builder_1.SarifRuleBuilder().initSimple({
                    ruleId: rule.id,
                    shortDescriptionText: rule.description,
                    helpUri: rule.link,
                });
                sarifRunBuilder.addRule(sarifRuleBuilder);
            });
        });
        arrAllMessages.forEach((result) => {
            result.messages.forEach((message) => {
                const sarifResultBuilder = new node_sarif_builder_1.SarifResultBuilder();
                const ruleId = message.rule.id;
                const sarifResultInit = {
                    level: message.type === "info"
                        ? 'note'
                        : message.type.toString(),
                    messageText: message.message,
                    ruleId: ruleId,
                    fileUri: path
                        .relative(process.cwd(), result.file)
                        .replace(/\\/g, '/'),
                    startLine: message.line,
                    startColumn: message.col,
                    endLine: message.line,
                    endColumn: message.col,
                };
                sarifResultBuilder.initSimple(sarifResultInit);
                sarifRunBuilder.addResult(sarifResultBuilder);
            });
        });
        sarifBuilder.addRun(sarifRunBuilder);
        const sarifContent = sarifBuilder.buildSarifJsonString({ indent: true });
        try {
            const sarifJson = JSON.parse(sarifContent);
            const rules = sarifJson.runs[0].tool.driver.rules;
            rules.forEach((rule) => {
                const ruleMarkdown = getRuleMarkdown(rule.id);
                if (ruleMarkdown) {
                    rule.help = {
                        text: rule.shortDescription.text,
                        markdown: ruleMarkdown,
                    };
                }
            });
            const updatedSarifContent = JSON.stringify(sarifJson, null, 2);
            (0, fs_1.writeFileSync)('htmlhint.sarif', updatedSarifContent);
            console.log(updatedSarifContent);
        }
        catch (error) {
            (0, fs_1.writeFileSync)('htmlhint.sarif', sarifContent);
            console.log(sarifContent);
        }
    });
};
module.exports = sarifFormatter;
